jax/random.py

```diff
@@ -564,25 +565,153 @@ def _cond_fn(kXVU):
                            lax.ge(lax.log(U), lax.add(lax.mul(X, one_over_two),
                                                       lax.mul(d, lax.add(lax.sub(one, V),
                                                                          lax.log(V))))))
-    return lax.bitwise_or(lax.le(V, zero), cond)
+    return cond
 
   def _body_fn(kXVU):
     key = kXVU[0]
-    key, x_key, U_key = split(key, 3)
-    x = normal(x_key, (), dtype=dtype)
-    v = lax.add(one, lax.mul(x, c))
+
+    def _next_kxv(kxv):
+      x_key = kxv[0]
+      x = normal(x_key, (), dtype=dtype)
+      v = lax.add(one, lax.mul(x, c))
+      k, = split(x_key, 1)
+      return k, x, v
+
+    key, x, v = lax.while_loop(lambda kxv: lax.le(kxv[2], zero), _next_kxv, (key, zero, minus_one))
     X = lax.mul(x, x)
     V = lax.mul(lax.mul(v, v), v)
-    U = uniform(U_key, (), dtype=dtype)
+    U = uniform(key, (), dtype=dtype)
+    key, = split(key, 1)
     return key, X, V, U
 
   # initial state is chosen such that _cond_fn will return True
-  _, _, V, _ = lax.while_loop(
-      _cond_fn, _body_fn, (key, zero, _constant_like(alpha, -1), zero))
+  _, _, V, _ = lax.while_loop(_cond_fn, _body_fn, (key, zero, one, _constant_like(alpha, 2)))
   z = lax.mul(lax.mul(d, V), boost)
   return lax.select(lax.eq(z, zero), onp.finfo(z.dtype).tiny, z)
 
 
+_bivariate_coef = [[0.16009398, -0.094634816, 0.025146379, -0.0030648348,
+                    1, 0.3266811, 0.10406087, 0.0014179033],
+                   [0.53487893, 0.12980707, 0.06573594, -0.0015649787,
+                    0.16639465, 0.020070098, -0.0035938937, -0.00058392601],
+                   [0.040121005, -0.0065914079, -0.002628604, -0.0013441777,
+                    0.017050642, -0.0021309345, 0.00085092385, -1.5248239e-07]]
+
+
+def _gamma_grad_one(z, alpha):
+    # Ref 1: Pathwise Derivatives Beyond the Reparameterization Trick, Martin & Fritz
+    # Ref 2: Case 4 follows https://github.com/fritzo/notebooks/blob/master/gamma-reparameterized.ipynb
+
+    # TODO: use lax.cond instead of lax.while_loop when available
```

Comment:
Should we make a feature request issue for what's needed here?

Reply:
@mattjj This feature is reported at https://github.com/google/jax/issues/490.

Reply:
Excellent, thanks!